controller/run.go

```diff
@@ -74,3 +94,70 @@ func RunManaged(cfg config.ShipperConfig) error {
 
 	return err
 }
+
+// runController is the main runloop for the shipper itself, and managed communication with the agent.
+func runController(ctx context.Context, agentClient client.V2) error {
+	log := logp.L()
+
+	err := agentClient.Start(ctx)
+	if err != nil {
+		return fmt.Errorf("error starting connection to client")
+	}
+
+	log.Debugf("Starting error reporter")
+	go reportErrors(ctx, agentClient)
+
+	log.Debugf("Started client, waiting")
+	handler := newClientHandler()
+
+	sigc := make(chan os.Signal, 1)
+	signal.Notify(sigc, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)
+
+	// use a timer to "bounce" the shipper based on external changes
+	// useful since we require two units for the shipper to actually start, which can arrive at different times, or get updated with different states
+	// This also makes it a little easier to deal with multple events requiring
+	// a (re)start of the shipper backend, in which case we'll only restart once with the latest config.
+	bounceTime := 200 * time.Millisecond // give us enough time to get multiple units before the shipper takes any action
```

Comment:
Inspired by the similar implementation in Beats I assume? 

https://github.com/elastic/beats/blob/9754d2b04f661e797cae113116785509d7d5ec89/x-pack/libbeat/management/managerV2.go#L335-L386

In that case it was necessary to group unit changes together because we have to pass the entire Beat configuration to the underlying config reloader, so it's diff algorithm for what was added and removed can keep working.

Is the same necessary here? Is there any reason why the gRPC server and the output itself can't be configured independently? Why do we need this logic to group unit changes together here?

Reply:
So, my thinking was that we don't really _want_ them configured independently. Treating the gRPC endpoint and the rest of the shipper pipeline independently adds a handful of technical questions like "what do we do with received events when the gRPC server is up, but there's no queue or output?" and I wasn't sure if we really got any benefits for the added complexity. I figured that either way, we'll want to confirm the expected state and config values of both units before we do anything, so the simplest solution was to try to synchronize the events and start everything at once.

Reply:
> "what do we do with received events if we start the gRPC endpoint before we start the queue or output?"

I would just return the gRPC Unavailable status code if we can detect that the queue isn't initialized yet. The Beat shipper output is going to be created asynchronously from any of this initialization anyway, clients always need to handle the case where they tried to connect before the shipper process started. They can just retry the same as they would if they got the equivalent "connection refused" error because the server isn't started yet.

You can return this with something like `status.New(codes.Unavailable, "initializing")` or similar https://pkg.go.dev/google.golang.org/grpc/status#New

We may already handle this but I'm not sure if there is a test where the queue isn't created yet. Just blocking until an event is accepted into the not yet initialized queue also works for this.

https://github.com/cmacknz/elastic-agent-shipper/blob/9d1a866c8333c3da4c5f94fc66daeeb61411d255/server/server.go#L113-L117

Reply:
I don't think there is a need to add complexity to the unit handling for this when the client can just retry or block without consequence.

I also think we should be able to reconfigure the gRPC server and the queue+output independently. We shouldn't need to teardown or reconfigure the gRPC server because the output type changed or the queue size was changed.

Reply:
> I would just return the gRPC Unavailable status code if we can detect that the queue isn't initialized yet.

Ah, didn't think of that, or just assumed we wouldn't wanna do it.

> We shouldn't need to teardown or reconfigure the gRPC server because the output type changed or the queue size was changed.

If we're going to be returning an `Unavailable` code in these instances, is there a meaningful difference between that and just shutting gRPC server down? Do we just want to make the current state 

I'm gonna start tinkering with some kind of shim that'll let the server start before we have any kind of publisher interface, see how it goes.

Reply:
> If we're going to be returning an Unavailable code in these instances, is there a meaningful difference between that and just shutting gRPC server down? Do we just want to make the current state

It lets the client tell the difference between the shipper reconfiguring itself and being stopped or having crashed, which will probably avoid some questions coming our way as people start to develop against it more and notice "connection refused" errors in the logs after everything was running previously. It will be much better to return an explicit unavailable code with a "reconfiguring" message or similar in this case.

Reply:
Resolving this, since we've separated the components.